src/main/kotlin/org/rust/ide/refactoring/generateConstructor/GenerateConstructorHandler.kt

```diff
@@ -31,37 +31,59 @@ class GenerateConstructorAction : CodeInsightAction() {
+    override fun isValidFor(editor: Editor, file: PsiFile): Boolean = getContext(editor, file) != null
 
     override fun startInWriteAction() = false
-    private fun getStructItem(editor: Editor, file: PsiFile): RsStructItem? =
-        file.findElementAt(editor.caretModel.offset)?.ancestorOrSelf()
+    private fun getContext(editor: Editor, file: PsiFile): Context? {
+        val element = file.findElementAt(editor.caretModel.offset)?: return null
+        val struct = element.ancestorOrSelf<RsStructItem>()
+        return if (struct != null) {
+            Context(struct)
+        }
+        else {
+            val impl = element.ancestorOrSelf<RsImplItem>() ?: return null
+            if (!isImplBlockValid(impl)) return null
+
+            Context(impl.typeReference?.baseType?.path?.reference?.resolve() as? RsStructItem ?: return null, impl)
+        }
+    }
 
     override fun invoke(project: Project, editor: Editor, file: PsiFile) {
-        val structItem = getStructItem(editor, file) ?: return
-        generateConstructorBody(structItem, editor)
+        val context = getContext(editor, file) ?: return
+        generateConstructorBody(context, editor)
     }
 
-    private fun generateConstructorBody(structItem: RsStructItem, editor: Editor) {
+    private fun generateConstructorBody(context: Context, editor: Editor) {
         checkWriteAccessNotAllowed()
-        val chosenFields = showConstructorArgumentsChooser(structItem.project, structItem) ?: return
+        val chosenFields = showConstructorArgumentsChooser(context.struct.project, context.struct) ?: return
         runWriteAction {
-            insertNewConstructor(structItem, chosenFields, editor)
+            insertNewConstructor(context.struct, context.implBlock, chosenFields, editor)
         }
     }
 
-    private fun insertNewConstructor(structItem: RsStructItem, selectedFields: List<ConstructorArgument>, editor: Editor) {
+    private fun isImplBlockValid(impl: RsImplItem): Boolean {
+        if (impl.traitRef != null) return false
+
+        return impl.members?.functionList?.find { it.name == "new" } == null
+    }
+
+    private fun insertNewConstructor(structItem: RsStructItem, implBlock: RsImplItem?,
+                                     selectedFields: List<ConstructorArgument>, editor: Editor) {
         checkWriteAccessAllowed()
         val project = editor.project ?: return
         val structName = structItem.name ?: return
         val psiFactory = RsPsiFactory(project)
-        val impl = psiFactory.createInherentImplItem(structName, structItem.typeParameterList, structItem.whereClause)
+        val impl = implBlock ?: let {
+            val impl = psiFactory.createInherentImplItem(structName, structItem.typeParameterList, structItem.whereClause)
+            structItem.parent.addAfter(impl, structItem) as RsImplItem
+        }
         val anchor = impl.lastChild.lastChild
         val constructor = getFunction(structItem, selectedFields, psiFactory)
```

Comment:
Note, this code doesn't take into account type arguments because it wasn't needed in previous implementation, i.e. for the following code
```rust
struct S<T>(T);

impl S<i32> {
    /*caret*/
}
```
you generate invalid code
```rust
impl S<i32> {
    pub fn new(field0: T) -> Self {
        S(field0)
    }
}
```

Reply:
Thanks, I didn't think of that.